src/main/java/com/minecolonies/coremod/client/gui/WindowAssignCitizen.java

```diff
@@ -0,0 +1,203 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.minecolonies.api.util.constant.TranslationConstants.*;
+
+/**
+ * Window for the hiring or firing of a worker.
+ */
+public class WindowAssignCitizen extends Window implements ButtonHandler
+{
+    /**
+     * Id of the done button in the GUI.
+     */
+    private static final String BUTTON_DONE = "done";
+
+    /**
+     * Id of the cancel button in the GUI.
+     */
+    private static final String BUTTON_CANCEL = "cancel";
+
+    /**
+     * Id of the citizen name in the GUI.
+     */
+    private static final String CITIZEN_LABEL = "citizen";
+
+    /**
+     * Id of the citizen list in the GUI.
+     */
+    private static final String CITIZEN_LIST = "unassigned";
+
+    /**
+     * Assign button of the gui.
+     */
+    private static final String CITIZEN_DONE = "done";
+
+    /**
+     * Link to the xml file of the window.
+     */
+    private static final String BUILDING_NAME_RESOURCE_SUFFIX = ":gui/windowAssignCitizen.xml";
+
+    /**
+     * Id of the job label in the GUI.
+     */
+    private static final String CITIZEN_JOB = "job";
+
+    /**
+     * Contains all the citizens.
+     */
+    private List<CitizenDataView> citizens = new ArrayList<>();
+    /**
+     * The view of the current building.
+     */
+    private final AbstractBuilding.View building;
+
+    /**
+     * List of citizens which can be assigned.
+     */
+    private final ScrollingList citizenList;
+
+    /**
+     * The colony.
+     */
+    private final ColonyView colony;
+
+    /**
+     * Constructor for the window when the player wants to assign a worker for a certain home building.
+     *
+     * @param c          the colony view.
+     * @param buildingId the building position.
+     */
+    public WindowAssignCitizen(final ColonyView c, final BlockPos buildingId)
+    {
+        super(Constants.MOD_ID + BUILDING_NAME_RESOURCE_SUFFIX);
+        this.colony = c;
+        building = colony.getBuilding(buildingId);
+        citizenList = findPaneOfTypeByID(CITIZEN_LIST, ScrollingList.class);
+        updateCitizens();
+    }
+
+    /**
+     * Clears and resets/updates all citizens.
+     */
+    private void updateCitizens()
+    {
+        citizens.clear();
+        citizens.addAll(colony.getCitizens().values());
+
+        //Removes all citizens which already have a job.
+        citizens = colony.getCitizens().values().stream()
+                     .filter(citizen -> citizen.getHomeBuilding() == null)
+                     .collect(Collectors.toList());
+    }
+
+    /**
+     * Called when the GUI has been opened.
+     * Will fill the fields and lists.
+     */
+    @Override
+    public void onOpened()
+    {
+        updateCitizens();
+        citizenList.enable();
+        citizenList.show();
+        //Creates a dataProvider for the homeless citizenList.
+        citizenList.setDataProvider(new ScrollingList.DataProvider()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Reduce this anonymous class number of lines from 26 to at most 20, or make it a named class. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)